Sources/Confidence/EventStorage.swift

```diff
@@ -0,0 +1,104 @@
+import Foundation
+
+internal protocol EventStorage {
+    func rollover() throws
+    func writeEvent(event: Event) throws
+    func batchReadyFiles() throws -> [URL]
+    func eventsFrom(fileURL: URL) throws -> [Event]
+}
+
+internal class EventStorageImpl: EventStorage {
+    let DIRECTORY = "events"
+    let EVENT_WRITE_DELIMITER = ",\n"
+    let READY_TO_SEND_EXTENSION = ".ready"
+    let encoder = JSONEncoder()
+    let decoder = JSONDecoder()
+    var folderURL: URL = URL(string: "")!
+    var fileURL: URL = URL(string: "")!
+    var currentBatch: [Event] = []
+
+    init() throws {
+        folderURL = try getFolderURL()
+        try createEventsDirectory(url: folderURL)
+        fileURL = folderURL.appendingPathComponent("events-\(Date().currentTime)")
+    }
+
+    func rollover() throws {
+        let newPath = URL(string: "\(fileURL)" + "\(Date().currentTime)")!
+        try FileManager.default.moveItem(at: fileURL, to: newPath)
+        try FileManager.default.removeItem(at: fileURL)
+        fileURL = folderURL.appendingPathComponent("events-\(Date().currentTime)")
+        currentBatch = []
+    }
+    
+    func writeEvent(event: Event) throws {
+        currentBatch.append(event)
+        let data = try encoder.encode(currentBatch)
+        try data.write(to: fileURL, options: .withoutOverwriting)
+    }
+    
+    func batchReadyFiles() throws -> [URL] {
+        var readyFilesList: [URL] = []
+        let directoryContents = try FileManager.default.contentsOfDirectory(atPath: folderURL.absoluteString)
+        for file in directoryContents {
+            if file.hasSuffix(READY_TO_SEND_EXTENSION) {
+                readyFilesList.append(URL(string: file)!)
+            }
+        }
+        return readyFilesList
+    }
+    
+    func eventsFrom(fileURL: URL) throws -> [Event] {
+        let data = try Data(contentsOf: fileURL)
+        let events = try decoder.decode([Event].self, from: data)
+        return events
+    }
+
+    private func getFolderURL() throws -> URL {
+        let rootFolderURL = try FileManager.default.url(
+            for: .cachesDirectory,
+            in: .userDomainMask,
+            appropriateFor: nil,
+            create: false
+        )
+        var nestedFolderURL: URL
+        if #available(iOS 16.0, *) {
+            nestedFolderURL = rootFolderURL.appending(path: DIRECTORY)
+        } else {
+            nestedFolderURL = rootFolderURL.appendingPathComponent(DIRECTORY, isDirectory: true)
+        }
+        return nestedFolderURL
+    }
+
+    private func createEventsDirectory(url: URL) throws {
+        do {
+            try FileManager.default.createDirectory(
+                at: url,
+                withIntermediateDirectories: false
```

Comment:
why `false`?

Reply:
I've just read up about it a bit more; it looks like we can omit the whole "create directory if doesn't exist" by setting `withIntermediateDirectories` to `true` - in this case, it will automatically create all the directories that don't exist given a URL